### PR TITLE
NIFI-9281 Enable Building on Java 17

### PR DESCRIPTION
**Pull request by exceptionfactory**
#### Description of PR

NIFI-9281 Updates a number of tests and several dependencies to support building the project on Java 17. The majority of the changes involve adjustments to unit tests in order to avoid illegal reflective access, which throws an error by default on Java 17. 

Dependency updates include upgrading the Maven WAR Plugin, the QuestDB driver for NiFi Framework status history, and the version of Groovy used in several Graph Processors. The GitHub actions workflow includes a new build target for Java 17 on Ubuntu Linux.

Although these changes enable a build across the project, some extension components may require further changes for full runtime support. Additional changes for particular components can be addressed in subsequent pull requests.

The following list summarizes the changes:

- Added Ubuntu Zulu JDK 17 GitHub build
- Adjusted MiNiFi C2 FileSystemConfigurationCache test to avoid using environment variables
- Adjusted MiNiFi StatusLogger and StatusLoggerTest to avoid overriding private logger
- Adjusted failure reason attribute check in TestGetIgniteCache
- Adjusted TestRangerAuthorizer and TestRangerNiFiAuthorizer to avoid checking nested exceptions
- Adjusted encrypt-config TestUtil to avoid unnecessary comparison of different types
- Disabled Javascript tests on Java 15 and higher due to removal of Nashorn from the JRE
- Disabled several Hive 3 tests on Java 17 for StringInternUtils illegal access
- Refactored nifi-enrich-processors tests to use Mockito without Powermock
- Refactored nifi-toolkit-tls tests to avoid illegal reflective access
- Removed deprecated X509Certificate test in CertificateUtilsTest
- Removed kryo serialization from nifi-site-to-site-client test
- Updated TestHashContent to use SHA-1 instead of SHA for hash algorithm
- Upgraded maven-war-plugin from 2.5 to 3.3.2
- Upgraded nifi-graph-bundle dependencies from Groovy 2.5.14 to 3.0.8
- Upgraded QuestDB from 4.2.1 to 6.2.1 in nifi-framework-core

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by MikeThomsen** [comment]
I'll try to review this later tonight or tomorrow morning. Would love to have an excuse to go to Java 17 on the job :-D

**Review by greyp9 — COMMENTED** [review]
This all looks really solid.  I tackled this by building from HEAD using a fresh Java 17 install, and using the PR as a reference.  Had a couple of questions about fidelity of test cases, but nothing dire.

File: nifi-commons/nifi-site-to-site-client/src/test/java/org/apache/nifi/remote/client/socket/TestSiteToSiteClient.java [hunk 1]
```diff
@@ -16,89 +16,18 @@
...
-            assertEquals(clientConfig.getUrls(), clientConfig2.getUrls());
-            // Serialization works, but the state manager is not serialized.
-            assertNotNull(clientConfig.getStateManager());
-            assertNull(clientConfig2.getStateManager());
-        } finally {
-            input.close();
-        }
-    }
-
+    @SuppressWarnings("deprecation")
```
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-commons/nifi-site-to-site-client/src/test/java/org/apache/nifi/remote/client/socket/TestSiteToSiteClient.java:30]: The `kryo` dependency did not have any direct runtime references, as indicated by the test scope, so other than rewriting the tests using standard Java Obejct serialization, I'm not sure what would be considered a comparable approach.

File: nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/test/java/org/apache/nifi/processors/TestISPEnrichIP.java [hunk 2]
```diff
@@ -252,8 +240,6 @@ public void whenInetAddressThrowsUnknownHostFlowFileShouldBeSentToNotFound() thr
         final Map<String, String> attributes = new HashMap<>();
         attributes.put("ip", "somenonexistentdomain.comm");
 
-        when(InetAddress.getByName("somenonexistentdomain.comm")).thenThrow(UnknownHostException.class);
```
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/test/java/org/apache/nifi/processors/TestISPEnrichIP.java:255]: The purpose of this mock was to ensure that this lookup would always throw an `UnknownHostException`.  Unfortunately, static mocking can be problematic in some cases.  Removing the mock resulted in the expected exception being thrown.

File: nifi-nar-bundles/nifi-ranger-bundle/nifi-ranger-plugin/src/test/java/org/apache/nifi/ranger/authorization/TestRangerNiFiAuthorizer.java [hunk 3]
```diff
@@ -200,19 +181,7 @@ public void testKerberosEnabled() {
...
-        Throwable cause = e.getCause();
-        while (cause != null) {
-            if (cause instanceof LoginException) {
-                foundLoginException = true;
-                break;
-            }
-            cause = cause.getCause();
-        }
-        assertTrue(foundLoginException);
+        assertThrows(AuthorizerCreationException.class, () -> authorizer.onConfigured(configurationContext));
```
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-ranger-bundle/nifi-ranger-plugin/src/test/java/org/apache/nifi/ranger/authorization/TestRangerNiFiAuthorizer.java:184]: Yes, this does make the test looser in terms of expected exception checking. There appear to be other issues with Ranger and Java 17, so the other option is marking several of these methods as ignored on Java 17, but it seemed better to make the tests somewhat more general in this scenario.

File: nifi-toolkit/nifi-toolkit-tls/src/test/java/org/apache/nifi/toolkit/tls/util/TlsHelperTest.java [hunk 4]
```diff
@@ -108,37 +95,16 @@
...
     OutputStreamFactory outputStreamFactory;
 
-    @Rule
-    public TemporaryFolder tempFolder = new TemporaryFolder();
-
     private ByteArrayOutputStream tmpFileOutputStream;
 
     private File file;
 
-    private static void setUnlimitedCrypto(boolean value) {
```
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-toolkit/nifi-toolkit-tls/src/test/java/org/apache/nifi/toolkit/tls/util/TlsHelperTest.java:131]: Good question, this change implies that unlimited strength cryptography support is expected.  Java 8 Update 171 introduced this by default, so there should be a separate effort to remove all checks for unlimited strength cryptography.

File: nifi-registry/nifi-registry-extensions/nifi-registry-ranger/nifi-registry-ranger-plugin/src/test/java/org/apache/nifi/registry/ranger/TestRangerAuthorizer.java [hunk 5]
```diff
@@ -220,26 +204,7 @@ public void testKerberosEnabledWithoutKeytabOrPrincipal() {
...
-        Throwable cause = e.getCause();
-        while (cause != null) {
-            if (!(cause instanceof SecurityProviderCreationException)) {
-                foundOtherException = true;
-                break;
-            }
-            cause = cause.getCause();
-        }
-        assertFalse(foundOtherException);
+        assertThrows(SecurityProviderCreationException.class, () -> setup(registryProperties, mock(UserGroupProvider.class), configurationContext));
```
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-registry/nifi-registry-extensions/nifi-registry-ranger/nifi-registry-ranger-plugin/src/test/java/org/apache/nifi/registry/ranger/TestRangerAuthorizer.java:207]: As mentioned in relation to the other Ranger test changes, this does make the exception checking looser, trading off some precision for general test execution on all platforms.

File: nifi-toolkit/nifi-toolkit-tls/src/test/java/org/apache/nifi/toolkit/tls/util/TlsHelperTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (greyp9) [reply nifi-toolkit/nifi-toolkit-tls/src/test/java/org/apache/nifi/toolkit/tls/util/TlsHelperTest.java:131]: Yep, you're right!  I remember both policies being included in the distribution, but was fuzzy on the details.

https://bugs.openjdk.java.net/browse/JDK-8170157

**Review by greyp9 — APPROVED** [review]
Started up a dev instance (compile=17, runtime=17), and tested with a simple QueryDatabaseTableRecord -> PutDatabaseRecord flow; worked perfectly.

I imagine there will be runtime rough spots around the edges, but this is a solid start.  LGTM.

**Review by MikeThomsen — APPROVED** [review]

**Comment by MikeThomsen** [comment]
> I imagine there will be runtime rough spots around the edges, but this is a solid start. LGTM.

Same here. Going to merge because nothing jumped out at me as problematic with the changes, and it appears to run the whole test suite just fine w/ Java 17. Anything we find problematic going forward that's broken will probably be found through live trial and error.

**Pull request closed without merging**